actors/util/adt/array.go

```diff
@@ -81,8 +81,10 @@ func (a *Array) BatchDelete(ix []uint64) error {
 func (a *Array) ForEach(out runtime.CBORUnmarshaler, fn func(i int64) error) error {
 	return a.root.ForEach(a.store.Context(), func(k uint64, val *cbg.Deferred) error {
 		if out != nil {
-			// Why doesn't amt.ForEach() just return the value as bytes?
-			if err := out.UnmarshalCBOR(bytes.NewReader(val.Raw)); err != nil {
+			if deferred, ok := out.(*cbg.Deferred); ok {
+				// fast-path deferred -> deferred to avoid re-decoding.
+				*deferred = *val
```

Comment:
Ugh, I hate this deep internal knowledge requirement. This is not the only place I wrote that "Why doesn't..." comment for the same reason. 
- Could you find and improve the others too? 
- If it makes sense to you, could you file an issue on AMT to just return bytes?

Reply:
This case wasn't internal knowledge for once, just code sleuthing (ipfs doesn't use cbor-gen).

The crux of the matter is that CBOR objects are _not_ (byte) length prefixed. In order for the AMT to break the array of objects into individual objects, it needs to parse the CBOR structure of the objects (even though it just throws this structure away). This is what `Deferred` does.

If I were to guess why one _might_ want to pass in `cbg.Deferred` here, instead of a simple byte array, I'd guess:

1. It indicates that the object is a valid CBOR object.
2. In the future, we could probably memoize some of the information we learned in the initial decoding to speed up future work?

But those are just guesses. Honestly, `amt.ForEach` _should_ just take a pointer to the thing to decode and we shouldn't need this indirection. I'm not sure why it doesn't.

Reply:
By "internal" I just meant looking at the code inside HAMT, rather than behaviour described or implied by the methods and their docs.

Reply:
Oh, yeah, that...

Reply:
Moved into #781